### ISSUE TITLE
DFlash strips thinking tokens

### ISSUE THREAD

**Issue by deepsweet** [jundot/omlx#1068]
Following #866 – the issue is still there.

Tried on the latest main branch commit bac678ec72c97e497d05c3c6d637fa54f1b3d7e3:

```
Installed 2 packages in 1ms
 - dflash-mlx==0.1.3 (from git+https://github.com/jundot/dflash-mlx@814c4a1203744c96ee3aa00802cf495abdf77a59)
 + dflash-mlx==0.1.5.1 (from git+https://github.com/jundot/dflash-mlx@110f2611aeeb58df108689d75bfa4c25b81a8824)
 - omlx==0.3.8rc1 (from file:///.../omlx)
 + omlx==0.3.8 (from file:///.../omlx)
```

<img width="787" height="681" alt="Image" src="https://github.com/user-attachments/assets/fabd2938-eafd-40ba-867f-653564f711bb" />

**Comment by jundot** [comment]
Thanks for the repro. Confirmed and pinned the cause: DFlash bypasses the scheduler, so the `<think>\n` prefix that's normally prepended to the first chunk for reasoning models never runs, and the streaming ThinkingParser leaks the whole reasoning block into content. Pushed a fix (846356d) that ports `_detect_needs_think_prefix` from scheduler into the dflash engine. Ships in the next release.

**Comment by TomLucidor** [comment]
Would this same issue happen to Qwen3.5/Qwen3.6 and Gemma4 models as well?

**Comment by norbertpunk** [comment]
Still reproducible on v0.3.9 - 10.

It looks like DFlash doesn't detect the thinking stop token (`</think>`). When the thinking budget (4k) runs out, the model doesn't stop reasoning, instead thinking content overflows into the chat completion output, where it continues looping (repeating the same verification block dozens of times) until max_tokens is hit.

With DFlash disabled (same model, same prompt, same settings), the stop token is detected correctly, thinking stays within budget and the response is clean.

Test case: DFlash ON: ~15k tokens of leaked thinking loops in the visible output. DFlash OFF: ~500 tokens, clean separation between thinking and response. Both arrived at the correct answer.

Environment:
- oMLX v0.3.9 & v0.3.10
- macOS 15.7.7, M3 Max 64GB
- Model: Qwen3.6-35B-A3B (mlx-community 8bit)
- DFlash model: z-lab/Qwen3.6-35B-A3B-DFlash

**Comment by iceblove** [comment]
## Still reproducible on 0.6.1 — thinking budget never fires on the DFlash path

This is a follow-up to @norbertpunk's May 24 report: the "budget runs out but the model keeps thinking until max_tokens" half of this issue is **still unfixed in 0.6.1 (build 2323)**. The `<think>` prefix leak was fixed, but the thinking **budget** has no application point on the DFlash engine.

### Environment
- oMLX 0.6.1 (macOS app, build 2323), Apple M3 36GB unified memory
- Model: `underlotus/Qwen3.8-27B-heretic-ara-oQ4e-mtp` (VLM, `Qwen3_5ForConditionalGeneration`, vision tower intact) with DFlash draft `jfan/Qwen3.8-27B-heretic-dflash`
- Per-model settings: `thinking_budget_enabled: true`, `thinking_budget_tokens: 16401`, `max_tokens: 65536`, `dflash_enabled: true`, `reasoning_parser: "qwen_3_5"`

### Symptom
Request entered thinking mode and kept reasoning for **>1 hour** with no output. The 16,401-token budget never fired. Generation only stops when it approaches `max_tokens: 65536` (at ~15–20 tok/s that is 55–70 minutes, which matches exactly what we observed). Server log confirms the request was served by DFlash:

```
DFlashEngine loaded: target=Qwen3.8-27B-heretic-ara-oQ4e-mtp, draft=Qwen3.8-27B-heretic-dflash, max_ctx=unlimited, fallback=vlm, ...
```

`max_ctx=unlimited` means no fallback to the BatchedEngine/VLM engine ever happens.

### Root cause (read from 0.6.1 sources)
1. `ThinkingBudgetProcessor` (`omlx/api/thinking.py:391`) is only assembled in `Scheduler._build_sampler_and_processors` (`omlx/scheduler.py:5563`), guarded by `sampling_params.thinking_budget is not None` + `needs_think_prefix` / parser `thinking_end_text` + a resolvable `think_end_token_ids`.
2. `DFlashEngine.generate` (`omlx/engine/dflash.py:1351`) has **no `thinking_budget` parameter** — the signature only accepts `max_tokens, temperature, top_p, top_k, min_p, repetition_penalty, presence_penalty, stop`; `thinking_budget` travels in `**kwargs` and is silently dropped. The module docstring even states "DFlash bypasses the scheduler" (confirmed in `engine/dflash.py`).
3. `_detect_needs_think_prefix` was ported into the dflash engine (the #1068 fix), but the **budget processor was not** — there is no logits-processor application point in the DFlash event loop (`_stream_dflash_events` → `TokenEvent` handling).

### Expected behavior
Thinking budget should be enforced on the DFlash path, either by applying `ThinkingBudgetProcessor` (the snapshot/restore approach used for DSpark in #2617 seems applicable) or by falling back to the processor-capable engine for requests carrying `thinking_budget` (the strategy used for vlm_mtp in #2399).

Related: #2617 (DSpark counter drift, fixed), #2399 (vlm_mtp drops thinking_budget and other processors — same family of "speculative decode bypasses processors").
